plugins/out_s3/s3.c

```diff
@@ -1243,6 +1259,245 @@ static struct multipart_upload *create_upload(struct flb_s3 *ctx,
     return m_upload;
 }
 
+/* Adds an entry to upload queue */
+static void add_to_queue(struct flb_s3 *ctx, struct s3_file *upload_file,
+                 struct multipart_upload *m_upload_file, const char *tag, int tag_len)
+{
+    int ret;
+    struct upload_queue *upload_contents;
+    char *tag_cpy;
+
+    if (upload_file == NULL && m_upload_file == NULL) {
+        flb_plg_error(ctx->ins, "Tried to add entry without file to upload_queue");
+        flb_errno();
+        FLB_OUTPUT_RETURN(FLB_ERROR);
+    }
+
+    if (tag_len == 0 || strlen(tag) != tag_len) {
+        flb_plg_error(ctx->ins, "Tried to add entry with invalid tag");
+        flb_errno();
+        FLB_OUTPUT_RETURN(FLB_ERROR);
+    }
+
+    /* Create upload contents object and add to upload queue */
+    upload_contents = flb_malloc(sizeof(struct upload_queue));
+    if (upload_contents == NULL) {
+        flb_plg_error(ctx->ins, "Error allocating memory for upload_queue entry");
+        flb_errno();
+        FLB_OUTPUT_RETURN(FLB_ERROR);
+    }
+    upload_contents->upload_file = upload_file;
+    upload_contents->m_upload_file = m_upload_file;
+    upload_contents->tag_len = tag_len;
+    upload_contents->retry_counter = 0;
+    upload_contents->upload_time = -1;
+
+    /* Necessary to create separate string for tag to prevent corruption */
+    tag_cpy = flb_malloc(tag_len);
+    if (tag_cpy == NULL) {
+        flb_free(upload_contents);
+        flb_plg_error(ctx->ins, "Error allocating memory for tag in add_to_queue");
+        flb_errno();
+        FLB_OUTPUT_RETURN(FLB_ERROR);
+    }
+    strncpy(tag_cpy, tag, tag_len);
+    upload_contents->tag = tag_cpy;
+
+    /* Adding to mk_list is not thread safe, so mutex lock is necessary */
+    ret = pthread_mutex_trylock(&ctx->upload_queue_mutex);
+    if (ret != 0) {
+        flb_free(upload_contents);
+        flb_free(tag_cpy);
+        flb_plg_debug(ctx->ins, "Failed to lock mutex in add_to_queue");
+        FLB_OUTPUT_RETURN(FLB_ERROR);
+    }
+    mk_list_add(&upload_contents->_head, &ctx->upload_queue);
+    ret = pthread_mutex_unlock(&ctx->upload_queue_mutex);
+    if (ret) {
+        remove_from_queue(upload_contents);
+        flb_plg_error(ctx->ins, "Failed to unlock mutex for add_to_queue");
+        flb_errno();
+        FLB_OUTPUT_RETURN(FLB_ERROR);
+    }
+}
+
+/* Removes an entry from upload_queue */
+void remove_from_queue(struct upload_queue *entry)
+{
+    mk_list_del(&entry->_head);
+    flb_free(entry->tag);
+    flb_free(entry);
+    return;
+}
+
+/* Validity check for upload queue object */
+static int upload_queue_valid(struct upload_queue *upload_contents, time_t now,
+                              void *out_context)
+{
```

Comment:
is this code truly needed? It looks sketchy to me. It looks like you are writing code which is trying to check if your other code is buggy or not. Do these cases happen in normal usage? Should they ever happen?

Reply:
This check should never happen in normal usage. They are there to narrow down the issue in the case that something really wrong happens. I believe these errors might be triggered if multi threading is turned on?